File: solutions/cpp/Gaussian_elimination.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <cmath>
#include <deque>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::deque;
using std::endl;
using std::ostream_iterator;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

void Eliminate_rows(int i, int j, vector<deque<bool>>* B);

// @include
deque<bool> Gaussian_elimination(const vector<deque<bool>>& A,
                                 const deque<bool>& y) {
  vector<deque<bool>> B(A);
  for (int i = 0; i < B.size(); ++i) {
    B[i].push_back(y[i]);
  }

  for (int i = 0; i < B.size(); ++i) {
    // Find the coefficient starting with 1.
    int idx = i;
    for (int j = i + 1; j < B.size(); ++j) {
      if (B[j][i]) {
        idx = j;
        break;
      }
    }
    swap(B[i], B[idx]);

    // Perform elimination except i-th row.
    if (B[i][i]) {
      Eliminate_rows(i, i, &B);
    }
  }

  for (int i = B.size() - 1; i >= 0; --i) {
    if (B[i][i] == false) {
      bool have_coefficient = false;
      for (int j = i + 1; j < A.size(); ++j) {
        if (B[i][j]) {
          Eliminate_rows(i, j, &B);
          have_coefficient = true;
          swap(B[i], B[j]);  // row permutation.
          break;
        }
      }

      if (!have_coefficient && B[i].back()) {
        cout << "No solution." << endl;
        return {};
      }
    }
  }

  deque<bool> x;
  for (int i = 0; i < B.size(); ++i) {
    x.push_back(B[i].back());
  }
  return x;
}

void Eliminate_rows(int i, int j, vector<deque<bool>>* B) {
  // Use B[i] to eliminate other rows' entry j.
  for (int a = 0; a < B->size(); ++a) {
    if (i != a && (*B)[a][j]) {
      for (int b = 0; b < (*B)[i].size(); ++b) {
        (*B)[a][b] = (*B)[a][b] ^ (*B)[i][b];
      }
    }
  }
}
// @exclude

bool check_answer_with_solution(const vector<deque<bool>>& A,
                                const deque<bool>& b, const deque<bool>& x) {
  for (int i = 0; i < A.size(); ++i) {
    bool res = A[i][0] && x[0];
    for (int j = 1; j < A[i].size(); ++j) {
      res = res ^ (A[i][j] && x[j]);
    }

    if (res != b[i]) {
      return false;
    }
  }
  return true;
}

bool check_answer_no_solution(const vector<deque<bool>>& A,
                              const deque<bool>& b) {
  // Generate all possible combinations of x to test
  // there is no solution actually.
  for (int val = 0; val < (1 << b.size()); ++val) {
    deque<bool> x;
    int temp = val;
    for (int i = 0; i < b.size(); ++i) {
      x.push_back(temp & 1);
      temp >>= 1;
    }
    /*
    copy(x.begin(), x.end(), ostream_iterator<bool>(cout, " "));
    cout << endl;
    */
    assert(check_answer_with_solution(A, b, x) == false);
  }
  return true;
}

void rand_matrix(vector<deque<bool>>* A) {
  default_random_engine gen((random_device())());
  for (int i = 0; i < A->size(); ++i) {
    for (int j = 0; j < (*A)[i].size(); ++j) {
      uniform_int_distribution<int> zero_or_one(0, 1);
      (*A)[i][j] = zero_or_one(gen);
    }
  }
}

void rand_vec(deque<bool>* b) {
  default_random_engine gen((random_device())());
  for (int i = 0; i < b->size(); ++i) {
    uniform_int_distribution<int> zero_or_one(0, 1);
    (*b)[i] = zero_or_one(gen);
  }
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  // Predefined tests.
  vector<deque<bool>> A(4);
  A[0] = {false, false, false, true};
  A[1] = {false, false, false, true};
  A[2] = {false, true, true, true};
  A[3] = {true, false, false, false};
  deque<bool> b = {true, true, false, true};
  deque<bool> x = Gaussian_elimination(A, b);
  copy(x.begin(), x.end(), ostream_iterator<bool>(cout, " "));
  cout << endl;
  if (x.empty()) {  // no solution
    assert(check_answer_no_solution(A, b));
  } else {  // solution
    assert(check_answer_with_solution(A, b, x));
  }

  // Perform random tests below
  for (int times = 0; times < 10000; ++times) {
    int n;
    if (argc == 2) {
      n = atoi(argv[1]);
      cout << "n = " << n << endl;
    } else {
      uniform_int_distribution<int> dis(1, 16);
      n = dis(gen);
    }
    vector<deque<bool>> A(n, deque<bool>(n));
    rand_matrix(&A);
    deque<bool> b(n);
    rand_vec(&b);
    deque<bool> x = Gaussian_elimination(A, b);
    cout << "n = " << n << endl;
    cout << "A = \n";
    for (int i = 0; i < A.size(); ++i) {
      copy(A[i].begin(), A[i].end(), ostream_iterator<bool>(cout, " "));
      cout << endl;
    }
    cout << "b = \n";
    copy(b.begin(), b.end(), ostream_iterator<bool>(cout, " "));
    cout << endl << endl;
    if (x.empty()) {  // no solution
      assert(check_answer_no_solution(A, b));
    } else {  // solution
      cout << "x = \n";
      copy(x.begin(), x.end(), ostream_iterator<bool>(cout, " "));
      cout << endl << endl;
      assert(check_answer_with_solution(A, b, x));
    }
  }
  return 0;
}

File: static/solutions/cpp/java/src/main/java/com/epi/GaussianElimination.java
package com.epi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import static com.epi.utils.Utils.copy;
import static com.epi.utils.Utils.simplePrint;

public class GaussianElimination {
  // @include
  static boolean[] gaussianElimination(List<List<Boolean>> A, boolean[] y) {
    List<List<Boolean>> B = copy(A);
    for (int i = 0; i < B.size(); ++i) {
      B.get(i).add(y[i]);
    }

    for (int i = 0; i < B.size(); ++i) {
      // Find the coefficient starting with 1.
      int idx = i;
      for (int j = i + 1; j < B.size(); ++j) {
        if (B.get(j).get(i)) {
          idx = j;
          break;
        }
      }
      Collections.swap(B, i, idx);

      // Perform elimination except i-th row.
      if (B.get(i).get(i)) {
        eliminateRows(B, i, i);
      }
    }

    for (int i = B.size() - 1; i >= 0; --i) {
      if (!B.get(i).get(i)) {
        boolean haveCoefficient = false;
        for (int j = i + 1; j < A.size(); ++j) {
          if (B.get(i).get(j)) {
            eliminateRows(B, i, j);
            haveCoefficient = true;
            Collections.swap(B, i, j); // row permutation.
            break;
          }
        }

        if (!haveCoefficient && B.get(i).get(B.get(i).size() - 1)) {
          System.out.println("No solution");
          return new boolean[0];
        }
      }
    }

    boolean[] x = new boolean[B.size()];
    for (int i = 0; i < B.size(); ++i) {
      x[i] = B.get(i).get(B.get(i).size() - 1);
    }
    return x;
  }

  static void eliminateRows(List<List<Boolean>> B, int i, int j) {
    // Use B[i] to eliminate other rows' entry j.
    for (int a = 0; a < B.size(); ++a) {
      if (i != a && B.get(a).get(j)) {
        for (int b = 0; b < B.get(i).size(); ++b) {
          boolean set = B.get(a).get(b) ^ B.get(i).get(b);
          B.get(a).set(b, set);
        }
      }
    }
  }
  // @exclude

  static boolean checkAnswerWithSolution(List<List<Boolean>> A, boolean[] b,
                                         boolean[] x) {
    for (int i = 0; i < A.size(); ++i) {
      boolean res = A.get(i).get(0) && x[0];
      for (int j = 1; j < A.get(i).size(); ++j) {
        res = res ^ (A.get(i).get(j) && x[j]);
      }

      if (res != b[i]) {
        return false;
      }
    }
    return true;
  }

  static boolean checkAnswerNoSolution(List<List<Boolean>> A, boolean[] b) {
    // Generate all possible combinations of x to test
    // there is no solution actually.
    for (int val = 0; val < (1 << b.length); ++val) {
      boolean[] x = new boolean[b.length];
      int temp = val;
      for (int i = 0; i < b.length; ++i) {
        x[i] = (temp & 1) > 0;
        // clang-format off
        temp >>>= 1;
        // clang-format on
      }

      /*
       * simplePrint(x); System.out.println();
       */

      assert !checkAnswerWithSolution(A, b, x);
    }
    return true;
  }

  static List<List<Boolean>> randMatrix(int m, int n) {
    Random gen = new Random();

    List<List<Boolean>> A = new ArrayList<>(m);
    List<Boolean> row;
    for (int i = 1; i <= m; i++) {
      row = new ArrayList<>(n);

      for (int j = 1; j <= n; j++) {
        row.add(gen.nextBoolean());
      }

      A.add(row);
    }

    return A;
  }

  static boolean[] randVec(int n) {
    boolean[] b = new boolean[n];
    Random gen = new Random();

    for (int i = 0; i < n; ++i) {
      b[i] = gen.nextBoolean();
    }

    return b;
  }

  public static void main(String[] args) {
    Random gen = new Random();

    // Predefined tests.
    List<List<Boolean>> A = new ArrayList<>(4);
    // java.util.Arrays.ArrayList<T> used in java.util.Arrays.asList(T...)
    // doesn't support adding a new element, so we'll wrap it with
    // java.util.ArrayList<E>
    A.add(new ArrayList<>(Arrays.asList(false, false, false, true)));
    A.add(new ArrayList<>(Arrays.asList(false, false, false, true)));
    A.add(new ArrayList<>(Arrays.asList(false, true, true, true)));
    A.add(new ArrayList<>(Arrays.asList(true, false, false, false)));

    boolean[] b = {true, true, false, true};

    boolean[] x = gaussianElimination(A, b);

    simplePrint(x);
    System.out.println();

    if (x.length == 0) { // no solution
      assert checkAnswerNoSolution(A, b);
    } else { // solution
      assert checkAnswerWithSolution(A, b, x);
    }

    // Perform random tests below
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
        System.out.println("n = " + n);
      } else {
        n = gen.nextInt(16) + 1;
      }

      A = randMatrix(n, n);

      b = randVec(n);

      x = gaussianElimination(A, b);
      System.out.println("n = " + n);
      System.out.println("A = ");

      for (List<Boolean> row : A) {
        simplePrint(row);
        System.out.println();
      }

      System.out.println("b = ");
      simplePrint(b);
      System.out.println();
      System.out.println();

      if (x.length == 0) { // no solution
        assert checkAnswerNoSolution(A, b);
      } else { // solution
        System.out.println("x = ");
        simplePrint(x);
        System.out.println();
        System.out.println();
        assert checkAnswerWithSolution(A, b, x);
      }
    }
  }
}
